packages/foundry/contracts/prediction-market/lib/PredictionMarketLib.sol

```diff
@@ -0,0 +1,297 @@
+     * @return bullAmount Bull units credited
+     * @return bearAmount Bear units credited 
+     */
+    function _addSingleSidedLiquidity(
+        PredictionMarket memory self,
+        Side side,
+        uint256 amount,
+        uint256 feePercentage
+    ) private returns (uint256 bullAmount, uint256 bearAmount) {
+        // Revert if there is no liquidity in the market, then amounts will be equal to the deposited amount
+        // otherwise amounts will equal the proportioned amount based on current market values
+        if(self.liquidity == 0){
+            (bullAmount, bearAmount) = side == Side.Bull ? (amount, uint256(0)) : (0, amount);
+        } else {
+            (uint256 quoteBull, uint256 quoteBear) = quote(self);
+            
+            bullAmount = side == Side.Bull ? Math.mulDiv(amount, 1e18, quoteBull) : 0;
+            bearAmount = side == Side.Bear ? Math.mulDiv(amount, 1e18, quoteBear) : 0;
+        }
+
+        // adjust the market balances based on the calculated values
+        self.liquidity += amount;
+        self.balanceBear += bearAmount;
+        self.balanceBull += bullAmount;
+        self.fees += _calculateFees(amount, feePercentage);
+    }
+
+    /**
+     * @notice Given an input amount and side, returns the maximum output amount of the other side
+     * @param self The prediction market
+     * @param amountIn The deposited liquidity amount
+     * @param side Side to swap FROM
+     * @param feeAmount Fee percentage to charge for the swap
+     * @return amountOut Output amount
+     * @return fees The fee amount collected for the swap
+     */
+    function getAmountOut(
+        PredictionMarket memory self,
+        Side side,
+        uint256 amountIn,
+        uint256 feeAmount
+    ) internal pure returns (uint256 amountOut, uint256 fees) {
+        (uint256 reserveIn, uint256 reserveOut) = side == Side.Bull ? 
+            (self.balanceBull, self.balanceBear) :
+            (self.balanceBear, self.balanceBull);
+
+        uint256 amountInAfterFees = Math.mulDiv(amountIn, 1e6-feeAmount, 1e6);
+        uint256 denominator = reserveIn * 1e6;
+
+        amountOut = Math.mulDiv(amountInAfterFees, reserveOut, denominator);
+        fees = amountIn - amountInAfterFees;
+    }
+
+    /**
+     * @notice Swap the given input amount and side from one side to the other
+     * @dev Swap fees are accumulated and deducted during market settlement
+     * @param self The prediction market
+     * @param amountIn The deposited liquidity amount
+     * @param side Side to swap FROM
+     * @param feeAmount Fee percentage to charge for the swap
+     * @return amountOut Units in the opposing side quoted from the swap
+     */
+    function swap(
+        PredictionMarket memory self,
+        Side side,
+        uint256 amountIn,
+        uint256 feeAmount
+    ) internal view returns (uint256 amountOut) {
+        // revert if the market has 0 liquidity on either side
+        if(!canSwap(self)){
```

Comment:
Nit; we test for negatives explicitly for clarity; like `canSwap(self) == false`

Reply:
👍 make sense